### PR TITLE
Make the search box filter apps

### PR DESCRIPTION
**Pull request by icewind1991**
Makes the search box filter apps in the app manager.

searches both name and description and sorts the description matches above the description matches

cc @DeepDiver1975 @jancborchardt @butonic

### CODE REVIEW

**Comment by LukasReschke** [comment]
Image is flickering when typing something into the search box – but otherwise works fine :+1:

File: settings/js/apps.js [hunk 1]
```diff
@@ -330,11 +330,40 @@ OC.Settings.Apps = OC.Settings.Apps || {
...
+		});
+
+		apps = apps.concat(_.filter(OC.Settings.Apps.State.apps, function (app) {
+			return app.description.toLowerCase().indexOf(query) !== -1;
+		}));
+
+		apps = _.uniq(apps, function(app){return app.id});
+
+		var source = $("#app-template").html();
+		var template = Handlebars.compile(source);
```
Comment (MorrisJobke) [inline settings/js/apps.js]: Maybe compile this once and cache it?

File: settings/js/apps.js [hunk 2]
```diff
@@ -330,11 +330,40 @@ OC.Settings.Apps = OC.Settings.Apps || {
...
 			.hide()
 			.text('');
+	},
+
+	filter: function(query) {
+		query = query.toLowerCase();
+		$('#apps-list').html('');
+
+		var apps = _.filter(OC.Settings.Apps.State.apps, function (app) {
+			console.log(app.name);
```
Comment (MorrisJobke) [inline settings/js/apps.js]: Debug stuff?

**Comment by jancborchardt** [comment]
Nice, works well! Just two possible enhancements:

1. As @LukasReschke mentioned, the app image flickers. Is not breaking but could be improved.
2. Would be even better to filter all apps instead of only the current view. Prioritized by the currently selected view, but still showing the rest below. That way it’s a modeless search and you are more likely to get what you’re looking for.

**Comment by DeepDiver1975** [comment]
> Would be even better to filter all apps instead of only the current view. Prioritized by the currently selected view, but still showing the rest below. That way it’s a modeless search and you are more likely to get what you’re looking for.

That would be a bigger change - let's go for this approach - better then nothing ....

.... fix for the flicker is being pushed ...

**Comment by DeepDiver1975** [comment]
failing unit tests are unrelated

**Comment by jancborchardt** [comment]
Can we get the commits squashed please? ;)

**Comment by jancborchardt** [comment]
The »No search result in other places« is not really needed it seems. And squash. Otherwise :+1:

**Comment by PVince81** [comment]
Can you add a JS unit test for that "filter" function ? Maybe create a new "appSpec.js" file (and add it to "tests/karma.config.js")

**Comment by DeepDiver1975** [comment]
> The »No search result in other places« is not really needed it seems. And squash. Otherwise :+1:

both fixed now

@icewind1991 any news on the js unit tests?

**Comment by icewind1991** [comment]
> any news on the js unit tests?

Had a look at it but there doesn't seem to be any existing tests for the app managment nor does the code seem particular suitable for testing (all static) and I dont have enough experience with js testing to start from scratch

**Comment by PVince81** [comment]
Okay, I'll take a look then.

Just noticed that you had a blast using underscore JS :smile:

**Comment by PVince81** [comment]
Note: static classes are easy to test in Javascript. One can use `sinon.stub()` to stub the static methods.

**Comment by PVince81** [comment]
Done: https://github.com/owncloud/core/commit/e144d3aa49a1e4a3bb487ada9d83a2b4c81e6630?w=1

I had to move the initialization block into a method and use the "if window.TESTING" hack for now.

**Pull request merged**